app/main/api/index.ts

```diff
@@ -0,0 +1,71 @@
+import * as ipc from "app/common/ipc-protocol"
+import { Event } from "app/main/synthetic"
+import { JsonC } from "app/main/system"
+import { Routes, matchRoute } from "./routes"
+
+export { routes } from "./routes"
+
+/**
+ * Factory function that given a system and routes returns an ipc-main listener that dispatches
+ * events to handlers and sends the response of those handlers asynchronously.
+ *
+ * @param {T} system A system containing at least a JsonC component in order to (de)serialize the
+ * data received.
+ * @param {Routes<T>} routes A map of routes that pairs method names to handler functions of
+ * type Handler<T>.
+ * @returns {(event: Event, req: string) => void} The listener function
+ */
+export const asyncListenerFactory = genericListenerFactory(
+  (event, response) => {
+    event.sender.send(response)
+  }
+)
+
+/**
+ * Factory function that given a system and routes returns an ipc-main listener that dispatches
+ * events to handlers and sends the response of those handlers synchronously.
+ *
+ * @param {T} system A system containing at least a JsonC component in order
+ * to (de)serialize the data received.
+ * @param {Routes<T>} routes A map of routes that pairs method names to handler functions of type
+ * Handler<T extends JsonC>.
+ * @returns {(event: Event, req: string) => void} The listener function
+ */
+export const syncListenerFactory = genericListenerFactory(
+  (event, response) => {
+    event.returnValue = response
+  }
+)
+
+/**
+ * Helper function that given a event-responding function generates an ipc-main listener
+ * function that dispatches events to handlers and sends the response of those handlers
+ * using the given event-responding function.
+ *
+ * @param {(event: Event, response: string) => void} f The event-responding function.
+ * @returns {(system: T, routes: Routes<T>) => (event: Event, req: string) => void} The listener
+ * factory function.
+ */
+function genericListenerFactory(f: (event: Event, response: string) => void) {
+  return (system: JsonC, routes: Routes<JsonC>) => {
+    return async (event: Event, rawRequest: string) => {
+      let response
+      const json = system.json
+
+      try {
+        const rawRequestObject = await json.decode(rawRequest)
+        const request = await ipc.decodeRequest(rawRequestObject)
+        const { method, id } = request
+        const meta = { id, method }
+        const handler = await matchRoute(routes, method)
+        const result = await handler(system, request.data)
+        response = ipc.okResponse(result, meta)
+      } catch (e) {
+        response = ipc.errorResponse(e, { request: rawRequest })
+      }
+      const responseEncoded = await json.encode(response)
+
+      f(event, responseEncoded)
+    }
+  }
+}
```

Comment:
That was my first implementation but we have a linter rule that says that promise-returning functions must be async :/

Reply:
AFAICT it's already async, isn't it?

Reply:
If we want errors inside handlers not to turn into `errorResponse`s, the simplest thing we can do is just catching them and using them as results for our `okResponse`s:
```ts
.then(([request, handler]) => [request, handler(system, request.data).catch((e) => e)])
```

In this case, it may be necessary to define a minimal type which those errors should comply with, as otherwise it would be easy for someone writing renderer process code to think that the method always suceeds and totally forget about implementing error control.

I mean, the frontend API code must always expect a `Promise<Either<Error, whatever>>` that is only `reject`ed upon protocol or system exceptions and needs to deal with the `Either`.